typing/ctype.ml

```diff
@@ -2521,8 +2537,16 @@ let add_gadt_equation uenv source destination =
     let expansion_scope =
       Int.max (Path.scope source) (get_equations_scope uenv)
     in
+    let type_origin =
+      match Env.find_type source env with
+      | decl -> type_origin decl
+      | exception Not_found -> assert false
+    in
     let decl =
-      new_local_type ~manifest_and_scope:(destination, expansion_scope) () in
+      new_local_type
+        ~manifest_and_scope:(destination, expansion_scope)
+        type_origin
+    in
```

Comment:
I don't understand why we need to reuse `type_origin` here. This comes in play when I am turning an abstract type `type a` into an abbreviation `type a = b` where `b` is itself abstract. Why is it necessary to give `a` the same origin as `b` in this case? I would assume that the abbreviation will be expanded in the contexts than need it, and then the type-checker will work with `b` and access its origin directly.

Reply:
We are giving `type a = b` the same origin as the original `type a` definition, not the origin of `b`.

Reply:
Ah, but in this context we know already that `a` is abstract, so we don't really need the funny `type_origin : declaration -> type_origin` function.

Looking at the code I understand that it would be invasive to collect the origin at the point where we check that `a` is abstract (I think this is in `is_instantiable env path` called by the callers of `add_gadt_equation`, let's wait for view patterns before we turn it into a data-returning function), but maybe a local match that asserts that the type is abstract would be simpler and clearer?

Reply:
That function isn't "funny": all type declarations have an origin, but only abstract declarations have an interesting one. If we did try to add equations to types with constructors in the future, then we would still want to propagate their (boring) origins.

Reply:
I still find it funny, but you know the code better than I do so I will bet that you are more in the right. Resolved.